lfs: extract `DiffIndexScanner`

ttaylorr opened the pull request:
This pull request extracts a `DiffIndexScanner` type that will be used by the `status` command to yield more helpful information about why certain files are modified.

Since both the proposed changes to the status command _and_ the `ScanIndex()` operation of the new gitscanner require scanning the output of the `diff-index` command, this PR is only mean to extract that common functionality.

---

/cc @git-lfs/core

ttaylorr commented on lfs/diff_index_scanner.go:
```diff
@@ -0,0 +1,191 @@
+//
+// More information about each of its valid instances can be found:
+// https://git-scm.com/docs/git-diff-index
+type Status rune
```
@technoweenie I think this needs a better name, but I'm not sure what to call it. Exporting this field is necessary, since it'll be used from the `commands` package, but I think adding `lfs.Status` to the public API is too generic.

I was originally thinking of renaming this to be called `lfs.DiffIndexStatus` or `lfs.DiffIndexEntryStatus`. That got me thinking, what about moving these diff index changes to the `git` package, or a new `git/diffindex` package?

There are a few helpers that we'd have to move, like `lfs.startCommand`, but I don't think that'd be a problem.

What are your thoughts?

technoweenie replied:
I'm cool with `lfs.DiffIndexStatus` for now. My rough plan was to move all gitscanner stuff to a separate package once all the refactoring is done. I don't think a `git/diffindex` package is necessary right now.

ttaylorr commented on lfs/gitscanner_index.go:
```diff
@@ -106,67 +105,46 @@ func scanIndex(cb GitScannerFoundPointer, ref string) error {
-				if status == "M" {
-					sha1 = description[2] // This one is modified but not added
-				}
+			indexMap.Add(sha, &indexFile{
```
This `*indexFile` is kept over the new `DiffIndexEntry` type, since the `Name` field (and `sha` value) aren't entirely the output from the `diff-index` command.

- `Name` refers to either the new name of the file (if renamed), or the `SrcName` field.
- `Sha` refers to either the `DstSha`, or the `SrcSha`, if the status is 'M' for modified.

ttaylorr commented on lfs/diff_index_scanner.go:
```diff
@@ -0,0 +1,191 @@
+func (s *DiffIndexScanner) scan(line string) (*DiffIndexEntry, error) {
+	parts := strings.Split(line, "\t")
+	if len(parts) < 2 {
+		return nil, errors.Errorf("invalid line: %s", line)
```
These errors are also generic, and could potentially be extracted into types. I think this is a bit overkill for the `lfs` package (think: `lfs.DiffIndexScannerInvalidLineErr` 😅 ), but I think would be appropriate for a new package entirely.

That being said, their messages are too generic, and should at least be wrapped with `diff-index: `. I think the most appropriate place to do this would be in the public `Scan()` function from which this function is called.

technoweenie replied:
What about: `Invalid diff-index output: %s`

I don't think a new error type is necessary just have one. Introduce a common error type or interface if the caller needs to do something special with these errors.

ttaylorr commented on lfs/diff_index_scanner.go:
```diff
@@ -0,0 +1,191 @@
+// If any error was encountered in starting the command or closing its `stdin`,
+// that error will be returned immediately. Otherwise, a `*DiffIndexScanner`
+// will be returned with a `nil` error.
+func NewDiffIndexScanner(ref string, cached bool) (*DiffIndexScanner, *wrappedCmd, error) {
```
Last thing: the old `ScanIndex()` function had a `cmd.Wait()` after the `bufio.NewScanner(cmd.Stdout).Scan()` loop. I don't think that's necessary, since the cmd [will have already closed `exec/cmd.Stdout`][1] by the time that `Wait()` will have returned.

[1]: https://github.com/golang/go/blob/go1.8/src/os/exec/exec.go#L448

ttaylorr commented:
Ack, I lied in https://github.com/git-lfs/git-lfs/pull/2035#discussion_r106277999: removing `cmd.Wait()` makes the tests in `test/test-status.sh` race-y. I'll take a look now 👍 .

ttaylorr commented:
> Ack, I lied in #2035 (comment): removing `cmd.Wait()` makes the tests in `test/test-status.sh` race-y. I'll take a look now.

Two lies! What was causing the tests to fail was wrapping a `nil` error, causing the `errs` channel to be full and causing the command to exit without any information. I went ahead and fixed that in 42407fb.

_Now_ this should be ready to merge 😄

technoweenie commented on lfs/diff_index_scanner.go:
```diff
@@ -0,0 +1,194 @@
+
+	desc := strings.Fields(parts[0])
+	if len(desc) < 5 {
+		return nil, errors.Errorf("invalid description: %s", parts[0])
```
How about `invalid diff-index description: %s`?

technoweenie commented on lfs/diff_index_scanner.go:
```diff
@@ -0,0 +1,194 @@
+		SrcName: parts[1],
+	}
+
+	if score, err := strconv.Atoi(desc[4][1:]); err != nil {
```
I'm not sure of the chances, but nothing is really confirming that `desc[4]` has at least 1 character.

ttaylorr replied:
It's subtle, but using `strings.Fields` does guarantee that at each field is at least one, non-whitespace character long.

Since we check that the `len(desc)` is at least 5, we're guaranteed to have at least 5 non-empty slice elements.

[Here](https://play.golang.org/p/uvyIM5YDnt) is an example.

technoweenie commented on lfs/gitscanner_index.go:
```diff
@@ -106,67 +105,44 @@ func scanIndex(cb GitScannerFoundPointer, ref string) error {
 		for scanner.Scan() {
-			// Format is:
-			// :100644 100644 c5b3d83a7542255ec7856487baa5e83d65b1624c 9e82ac1b514be060945392291b5b3108c22f6fe3 M foo.gif
-			// :<old mode> <new mode> <old sha1> <new sha1> <status>\t<file name>[\t<file name>]
```
I like these old comments. Can you add them to the new `scan()` func somewhere?

ttaylorr replied:
Good call, I added those back in e16214c.

ttaylorr commented:
Thanks for the review, @technoweenie. I addressed your concern about `desc[4]` having no characters, and added the comments back from https://github.com/git-lfs/git-lfs/pull/2035#discussion_r106455809.

I think I may pass on rewording the error messages. I wrote them assuming that the caller would re-wrap them with `errors.Wrap(scanner.Error(), "diff-index")`, so the output would be:

- `diff-index: invalid line: ...`
- `diff-index: invalid description: ...`

Which I think is clear and not overly specific. Do you think this is ready to merge?

ttaylorr commented:
:wave: Thanks for the review!
